(#26) Fix tracking of unexecuted searchkick queries and tracking of queries across multiple indices

patrick-nits opened the pull request:
This pull request fixes #26

ankane commented:
Hey @patrick-nits, thanks for the PR! I'll leave a few comments inline.

ankane commented on app/controllers/searchjoy/searches_controller.rb:
```diff
@@ -39,7 +39,8 @@ def recent
     def set_search_types
-      @search_types = Searchjoy::Search.uniq.pluck(:search_type).sort
+      @search_types = Searchjoy::Search.uniq.pluck(:search_type)
+      @search_types = @search_types.reject(&:blank?).sort + @search_types.select(&:blank?)
```
Is there any reason to include `blank?` search types?

patrick-nits replied:
When using searches across **multiple indices** (with searchkick) there is no `name` nor `klass` available, thus the `search_type` will be blank. This results in displaying in an error when trying to access the searchjoy statistics panel. (Can´t do `[nil, 'String'].sort`)

ankane replied:
Is there a reason to do the above approach rather than

```ruby
Searchjoy::Search.uniq.pluck(:search_type).reject(&:blank?).sort
```

patrick-nits replied:
Well, that would exclude Searches with `search_type==nil`.

patrick-nits replied:
I dont know If it is _that_ important to include/ display Searches where the search_type is blank. Maybe I have mixed some requirements of my own project into this fix.

ankane commented on lib/searchjoy/track.rb:
```diff
@@ -5,11 +5,23 @@ def search_with_track(term, options = {}, &block)
       end
 
-      if options[:track]
+      if options[:track] && options[:execute]
```
Is this needed if we hook into the `execute` method instead? Passing `execute: true` will cause searches to be inserted twice.

patrick-nits replied:
Mh, good point. I guess hooking into `execute`will be sufficient. I´ll update the PR as soon as I can.

ankane commented on lib/searchjoy/track.rb:
```diff
@@ -5,11 +5,23 @@ def search_with_track(term, options = {}, &block)
+
+      if options[:track]
+        attributes = options[:track] == true ? {} : options[:track]
+        search_type = klass.try(:name) || options[:index_name].join(' ')
```
This throws an error if index_name is `nil` or not an array.

patrick-nits replied:
D´oh, you are right. I´ll fix this.

patrick-nits commented:
Hey @ankane, I updated the PR and removed the `search_with_track`method entirely. 
**Note**: The legacy reference in searchkick (searchkick-1.2.1/lib/searchkick/model.rb:2) for the module `Reindex` can be removed if this PR is merged.

ankane commented on lib/searchjoy/track.rb:
```diff
@@ -1,15 +1,22 @@
+        search_type = if klass.respond_to?(:name) && klass.name.present?
+                        klass.name
+                      elsif options[:index_name].is_a? Array
+                        options[:index_name].join(' ')
```
Would probably be good to use `index_name` if it's a string as well.

```ruby
Array(options[:index_name]).join(", ")
```

ankane commented on lib/searchjoy/track.rb:
```diff
@@ -1,15 +1,22 @@
+                      elsif options[:index_name].is_a? Array
+                        options[:index_name].join(' ')
+                      else
+                        nil
```
What if we had a default like `All Indices`?

ankane commented on lib/searchjoy.rb:
```diff
@@ -9,6 +9,7 @@
 module Searchjoy
   # time zone
   mattr_reader :time_zone
+
```
Can you remove the added whitespace lines to keep the diff minimal?

patrick-nits commented:
Updated to fix your comments.

ankane commented:
Great, thanks @patrick-nits 👍
